Generate sphinx docs with Docker

denzel-morris opened the pull request:
Remaining TODO:

- ~~Customize Sphinx CSS to fit our styling~~
- ~~Integrate Sphinx doc generation with Travis CI~~

I will ping this PR and update from WIP the above items are closed out.

denzel-morris commented:
### Generate Sphinx Docs
```bash
$ ./scripts/geneate_sphinx_docs.sh 2.0-dev
```
We use the `2.0-dev` branch for now because that's where the latest documentation exists. However, when no branch is passed, the script defaults to `master`.

This script requires the same version of `docker-ce` used by the main FS images. Once that is done, you can start the server like normal and navigate to `/docs/latest/develop/reference/index.html`.

### Travis CI Build Errors

There will be a few build errors on this branch that are not due to work done on this branch. Some Python files within the FS project need their docstrings updated to adhere to Google's Python Docstring style guide. Then Sphinx will be able to render their contents properly.

The three issues are:

- Spacing before code example/block
- Escaping `*` when not used at the start of a parameter name
- Broken links

The files that need their docstrings fixed can be inferred from the build errors on the documentation files:
```
google.cloud.forseti.common.gcp_api.bigquery.html
google.cloud.forseti.common.gcp_api.compute.html
google.cloud.forseti.common.gcp_api.container.html
google.cloud.forseti.common.util.regular_exp.html
google.cloud.forseti.enforcer.gce_firewall_enforcer.html
google.cloud.forseti.scanner.audit.iam_rules_engine.html
```

I can provide more details if necessary.

### Summary of Features

- Module Index categorizes modules while ignoring the common `google.cloud.forseti` prefix now
- Proper indentation and rendering for index and reference documentation pages
- All headers and functions support deep linking with clickable anchor (like other docs pages)
- All source code cross-referenced, linked, and highlighted; available under reference as well

mirons-google commented:
Thanks Denzel. Regarding the docstrings that need fixed would you mind going ahead and creating a PR to fix them?

denzel-morris commented:
Sure thing, absolutely, I will submit an additional PR this evening to
fixup those docstrings.

On Wed, Apr 18, 2018 at 12:10 PM, mirons <notifications@github.com> wrote:

> Thanks Denzel. Regarding the docstrings that need fixed would you mind
> going ahead and creating a PR to fix them?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/forseti-security/pull/1409#issuecomment-382442013>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABnY2PAYsgWlbVq0ctR0WTDtdv1hO6kbks5tp2V7gaJpZM4TTUEY>
> .
>

mirons-google commented on scripts/docker/data/conf.py:
```diff
@@ -0,0 +1,328 @@
+# -*- coding: utf-8 -*-
```
Can you please add the LICENSE here?

mirons-google commented on scripts/docker/data/conf.py:
```diff
@@ -0,0 +1,328 @@
+import sys
+import os
+
+sys.path.insert(0, os.path.abspath('.'))
```
@denzel-morris How do we maintain the inserts going forward w/o having to remember to do it? Can we wildcard this here or use a python method (os.walkdir)?

mirons-google commented on scripts/docker/data/conf.py:
```diff
@@ -0,0 +1,328 @@
+master_doc = 'index'
+
+# General information about the project.
+project = u'Forseti Security'
```
Can we pull this section of information from a canonical source?

mirons-google commented on scripts/docker/data/conf.py:
```diff
@@ -0,0 +1,328 @@
+# (source start file, target name, title,
+#  author, documentclass [howto, manual, or own class]).
+latex_documents = [
+    (master_doc, 'ForsetiSecurity.tex', u'Forseti Security Documentation',
```
Same here, can we pull this information from a canonical source?

mirons-google commented on scripts/docker/data/conf.py:
```diff
@@ -0,0 +1,328 @@
+# One entry per manual page. List of tuples
+# (source start file, name, description, authors, manual section).
+man_pages = [
+    (master_doc, 'forsetisecurity', u'Forseti Security Documentation',
```
Same comment as above.

mirons-google commented on scripts/docker/data/conf.py:
```diff
@@ -0,0 +1,328 @@
+# (source start file, target name, title, author,
+#  dir menu entry, description, category)
+texinfo_documents = [
+    (master_doc, 'ForsetiSecurity', u'Forseti Security Documentation',
```
Same comment as above.

mirons-google commented on scripts/docker/data/domainindex.html:
```diff
@@ -0,0 +1,43 @@
+    Template for domain indices (module index, ...).
+
+    :copyright: Copyright 2007-2016 by the Sphinx team, see AUTHORS.
+    :license: BSD, see LICENSE for details.
```
We'll need to put this into third_party. If moving the HTML file is too difficult at least create a sphinx entry with its LICENSE.

mirons-google commented on scripts/docker/data/genindex.html:
```diff
@@ -0,0 +1,75 @@
+    Template for an "all-in-one" index.
+
+    :copyright: Copyright 2007-2016 by the Sphinx team, see AUTHORS.
+    :license: BSD, see LICENSE for details.
```
Same comment regarding third_party and the LICENSE.

mirons-google commented on scripts/docker/data/defindex.html:
```diff
@@ -0,0 +1,40 @@
+    :copyright: Copyright 2007-2016 by the Sphinx team, see AUTHORS.
+    :license: BSD, see LICENSE for details.
+#}
+---
```
See the comment below regarding third_party and the LICENSE.

mirons-google commented on scripts/generate_sphinx_docs.sh:
```diff
@@ -0,0 +1,44 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+BUILD_FROM_BRANCH="${1:-master}"
```
Can you add comments here regarding what's happening on the various lines? Also please ensure it adheres to the Google style guide for [Bash](https://google.github.io/styleguide/shell.xml) if it doesn't.

mirons-google commented on travis/jekyll_build.sh:
```diff
@@ -15,6 +15,9 @@
 set -e
 trap 'return_code=$?' ERR
 
+# Write out the Python API documentation via Sphinx
```
Can you add a TODO here to change this after 2.0-dev becomes master.

mirons-google commented:
Friendly ping on where we're at with this PR @denzel-morris

denzel-morris commented:
Sorry for the latency, the last changeset addresses all outstanding items:

1. Add LICENSE to `scripts/docker/data/conf.py`
2. Recursively walk module directory in `conf.py` instead of explicitly naming each directory
3. Canonicalize the project name, author, and copyright by pulling project name from the `google.cloud.forseti.__package_name__` and deriving the author and copyright from the project name
4. Furthermore, removed the unnecessary master files and names for Latex and Man pages since we're not using any of that functionality
5. Added a directory in `third_party/` for `sphinx-doc` that contains the LICENSE along with the AUTHORS (since it's referenced by LICENSE) and a README that explains why we can't pull the HTML files directly from that directory
6. Refactored `generate_sphinx_docs.sh` to bring inline with Bash style guides and make the process more clear
7. Added a TODO tagged to me to switchover to `master` in `travis/jekyll_build.sh` when `2.0-dev` becomes `master`
